workflows/tests: run tap_syntax on stable `brew`

cho-m opened the pull request:
<!-- Use [x] to mark item done, or just click the checkboxes with device pointer -->

- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

carlocab reviewed (approved):
Love this.

Bo98 commented:
Looking at CI times, `matrix.branch` should be added to the style cache key and restore key I reckon

cho-m commented:
> Looking at CI times, `matrix.branch` should be added to the style cache key and restore key I reckon

Is it just:
```diff
-          key: style-cache-${{ github.sha }}
-          restore-keys: style-cache-
+          key: style-cache-${{ matrix.branch }}-${{ github.sha }}
+          restore-keys: style-cache-${{ matrix.branch }}-
```

Bo98 commented:
Yes. First build will be slow, but subsequent reruns should be quick again.

cho-m commented on .github/workflows/tests.yml:
```diff
@@ -40,13 +43,14 @@ jobs:
           core: true
           cask: false
           test-bot: true
+          stable: ${{ matrix.branch == 'stable' }}
```
I may need to adjust https://github.com/Homebrew/actions/blob/master/setup-homebrew/main.sh#L146 since we do have tags that aren't linked to `master` branch.

Bo98 replied:
Yeah 4.2.14 is a special out-of-band release so `master` being on 4.2.13 is intentional. Such scenarios are rare.

So that output is correct.

cho-m replied:
I was wondering if might miss latest tag in current action since we do `git fetch --force origin` rather than `git fetch --force --tags origin` (like in `brew` - https://github.com/Homebrew/brew/blob/master/Library/Homebrew/cmd/update.sh#L63)

Bo98 replied:
Oh right, that's probably because it's prefetched in a Docker container:

cho-m replied:
If there isn't much overhead, I can just add `--tags` to existing command or just run `git_retry -C "$HOMEBREW_REPOSITORY" fetch --force --tags origin` right before `git tag`.

Another option I considered was `brew update-reset "$HOMEBREW_REPOSITORY"` but didn't get chance to check on full impact of command.

Bo98 replied:
Adding `--tags` to the existing command makes sense

ZhongRuoyu commented:
Required check names need to be updated in repo settings. (Which also means all PRs need a rebase after this I think?)

MikeMcQuaid reviewed (approved):
Thanks again @cho-m, great idea!

cho-m commented:
Going to merge and try it out.

I updated workflow to try to keep existing required `tap_syntax` with same cache there. Changed variable to boolean `stable` to avoid specific master branch name (to reduce usage if we switch to main).

Action should work in current state for `homebrew-core`, but will open PR to grab `--tags` in case used elsewhere.

---

One scenario that could happen is if we want to introduce a new tap-wide audit/style/syntax that is incompatible between stable and master.

This could lead to `brew` PR failing on tap syntax of `homebrew-core` while a corresponding PR on this side would fail on `tap_syntax (stable)`.

We might have to bypass check and make sure to sync the merge of both PRs along with a new tagged release of `brew`. Or modify corresponding PRs to phase in the changes for master without impacting stable checks.

Maybe something along lines of:
```diff
--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -30,6 +30,10 @@ jobs:
     strategy:
       matrix:
         stable: [false, true]
+        new_syntax: contains(github.event.pull_request.labels.*.name, 'new syntax')
+        exclude:
+          - stable: true
+            new_syntax: true
     name: tap_syntax${{ matrix.stable && ' (stable)' || '' }}
     runs-on: ubuntu-22.04
     container:
```

Bo98 commented:
It would probably make sense to disable elements on the stable check. For example, `brew style` on stable probably doesn't make sense but `brew readall` does.

The label idea won't work as it'll affect every PR after merging.

cho-m commented:
> The label idea won't work as it'll affect every PR after merging.

Label would require new tag after whatever breaking syntax change is introduced. But this could be some ongoing development work to a new `brew` release so may not be possible to tag.

---

If there is a lot of conditional differences, may be easier to split matrix to separate jobs and selectively tune each.

Bo98 commented:
`brew style` changes are really never breaking syntax changes. Breaking changes should already caught by `brew readall` - if not then I consider that a bug.

cho-m commented:
> `brew style` changes are really never breaking syntax changes. Breaking changes should already caught by `brew readall` - if not then I consider that a bug.

Makes sense. Going to need changes in test-bot to tune this.

---

This did work for my PR - https://github.com/Homebrew/homebrew-core/actions/runs/8402834719/job/23012649264?pr=166501#step:5:20

* `brew style` only showed the non-impacting change.
* `brew readall` and `brew audit` both pointed to same breaking change.

MikeMcQuaid commented:
Agreed that ideally this would run `brew readall` only.
